### ISSUE TITLE
Generating pdf document from Sphinx project raises Package pdftex.def Error

### ISSUE THREAD

**Issue by pique-b** [sphinx-doc/sphinx#8135]
**Describe the bug**
Using command `make latexpdf` from Sphinx project's docs root raises the following **error**:

> ! Package pdftex.def Error: File `"""{img_file}".jpg' not found: using dra
ft setting.

**To Reproduce**

### **The following are the project's basic setup:**

**A.** In **_Conf.py_** (I am using _abspath_ as shown below):

```
sys.path.insert(0, os.path.abspath('..'))
...
...
# Options for LaTeX:
latex_engine = 'pdflatex'
latex_elements = {
    'papersize': 'a4paper',
    'figure_align': 'htbp',
    'pointsize': '10pt',
    
    # Additional stuff for the LaTeX preamble
    'preamble': r'''
        \usepackage{hyperref}
        \usepackage{graphicx}
        \usepackage{flafter}
        ''',
    }
```
**B.** The directory _tree_:

```
docs/           <= Sphinx Documentation folder 
    _build/
    _static/
        css/
        images/
           abc.jpg
        img_file.jpg      <= This image file is in the root ("_static") directory
    _templates/
    conf.py
    index.rst
    modules/              <= Folder containing my source (.rst) files
        execute/
            index.rst       <= Index within the sub-folder "execute"
            file_with_image.rst       <= I am trying to put the image in this source file
            some_other_file.rst
```

**C.** My _file_with_image.rst_ **file** where I am including the image:
```
**Adding an image**

Example of image usage using "image" directive:

.. image:: /_static/img_file.*
   :align: right
```
**D. The document flow:**
**_Root_** _index.rst_  **->** index.rst (in _**modules/execute/**_) **->** _file_with_image.rst_ 

**Command line that causes the error:**
```
$ make latexpdf
```

**Expected behavior**
The image included in **Step C** above, should get included in the generated pdf document.

**Environment info**
- OS: Mac 10.12.6
- Python version: 3.6.4
- Sphinx version: 3.1.2
- Sphinx extensions:  sphinx.ext.autodoc, imgmath, doctest, coverage, intersphinx, extlinks
- Extra tools: LaTeX

### **Additional context**
When the project's **.tex** file (in /_build/latex/ directory) is opened in a **Tex** editor, this is what the code fragment from the image insertion section shows:
```
\subsection{Add Image}
\label{\detokenize{modules/aud_execution/add_image_with_image_directive:add-image}}\label{\detokenize{modules/aud_execution/add_image_with_image_directive::doc}}
\sphinxstylestrong{Adding an image}

Example of image usage using “image” directive:

\noindent{\hspace*{\fill}\sphinxincludegraphics{{img_file}.jpg}}
```
Running the **QuickBuild** in the Tex editor raises error (as encountered in the terminal, described at the beginning of this document):

> ! Package pdftex.def Error: File `"""{img_file}".jpg' not found: ...

**However**, when the curly brackets from around the image file name, {{img_file}.jpg} are removed and changed it to simply `{img_file.jpg}` (**as shown below**):

`    \noindent{\hspace*{\fill}\sphinxincludegraphics{img_file.jpg}}`

**the document successfully compiles and the image can be seen in the pdf file.**

At this point, opening the pdf file in the **_build/taex/** subfolder shows the image as is expected.

**Comment by tk0miya** [comment]
Hmm... it's a bit strange. In my understanding, surrounding basename of file by braces is not wrong notation (see #2182). What version of pdflatex do you use?

**Comment by pique-b** [comment]
@tk0miya - Thanks for your reply. I am using pdfTeX 3.14159265-2.6-1.40.20 (TeX Live 2019) (I got the version using `latexpdf -v` in the terminal). Checked the referred issue #2182 but I think my situation is a bit different.

In the instant issue (#2182), the case was of multiple dots (.) in the file name and the solution was to kind of "mask" the file name up to the last dot (i.e. the **extension**). It seems a solution was built into **Sphinx** for LaTeX to work around the situation of multiple dots. This has got me thinking, **if that solution could be the reason why I am getting this additional braces around the image file name.** Or is it?

PS. Sorry for the late response.

**Comment by fleimgruber** [comment]
@tk0miya I can reproduce this (using latexmk and pdfTeX 3.14159265-2.6-1.40.20 (TeX Live 2019/W32TeX) as well). It works if the braces are removed. It seems that the issue might be with `\sphinxincludegraphics` and not with `\includegraphics` (#2182)?

**Comment by tk0miya** [comment]
@jfbu Could you check this please? I don't understand what happened in this case...

**Comment by jfbu** [comment]
@tk0miya I do not understand so far this case. 

 
>    ! Package pdftex.def Error: File `"""{img_file}".jpg' not found: using dra
>     ft setting.
> 

What is strange here are the quotes `"`. I can reproduce similar looking error message (without quotes) if the image file is not copied over to latex build repertory. (for example the image file is in `_static/` but the image directive was ``.. image:: img_file.*``). With the image present, and the tex file containing something such as

`\noindent\sphinxincludegraphics{{img_file2}.png}
`

There is no problem. There has been some changes in upstream LaTeX over the last 2 or 3 years in handling of graphics filenames. Indeed, in the past the extra brace pair `{..}` was need for filenames with multiple extensions, and recent LaTeX has improved handling, so it is not needed nowadays. But it doesn't hurt, I have no problem with

`\noindent\sphinxincludegraphics{{img_file2.png}.png}
`

With my TeXLive 2020 more or less up-to-date as of today, I can indeed remove the braces and it will still work. Quotes ``"`` have been used at some point in LaTeX upstream internally to handle graphics filenames with spaces. 

By the way I tested and

`\noindent\sphinxincludegraphics{{img file2.png}.png}
`

from

`.. image:: img\ file2.png.png
`

works fine. Modifying manually the latex into

`\noindent\sphinxincludegraphics{img file2.png.png}
`

i.e. a filename with space and double extension, and no extra braces, it still does work. But here I am sure it would have failed at LaTeX side some years ago.

What would be important here is the LaTeX format version, this is at top of the voluminous console output (from latex itself) of the latexmk run.  I am on Mac OS as well (an old one) with TeXLive 2020 (udpated some weeks ago) and near top of latexmk console output I get

```
This is pdfTeX, Version 3.14159265-2.6-1.40.21 (TeX Live 2020) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./test.tex
LaTeX2e <2020-10-01> patch level 2
L3 programming layer <2020-12-07> xparse <2020-03-03> (./sphinxmanual.cls
Document Class: sphinxmanual 2019/12/01 v2.3.0 Document class (Sphinx manual)

```

Also relevant is the actual `graphicx` latex package version (but on a normal tex install it is always updated at same time as the whole latex base stuff). Mine is

```
/usr/local/texlive/2020/texmf-dist/tex/latex/graphics/graphicx.sty
Package: graphicx 2020/09/09 v1.2b Enhanced LaTeX Graphics (DPC,SPQR)
```

which one sees in  the `.log` file in latex build repertory.

I am not sure here how `\sphinxincludegraphics` could be involved. All it does is invoking `\includegraphics` twice, the first time in draft mode in some temp box to get some dimensions, the second time with some options for width and height to avoid page overflows. Potentially the graphics filename could as it is handled there as a tex macro parameter be a bit altered, like handling a brace pair but this can not happen in `{name}.extension` situation. Anyway all of sphinx graphics inclusion has the brace pair and it works generally speaking.

Perhaps there was some unfortunate intermediate version of LaTeX/graphicx which is related to this issue but all I can reproduce is when the actual graphics files are not there in the latex build repertory. Else, with or without braces, with `\sphinxincludegraphics` or not, I can not reproduce so far.

**Comment by jfbu** [comment]
could this be related to https://github.com/latex3/latex2e/issues/204 ? @pique-b you mention using TL2019, is it a fully updated one ?

**Comment by jfbu** [comment]
@fleimgruber could you please check the file with extension `.log` inside the latex build repertory after a failed build to get the LaTeX format version (near top of file or console output) and also the graphicx package version (further down the file)? If LaTeX format does not  end in ``PL2`` (patch level 2)  or higher, then indeed there was a LaTeX upstream problem in fall 2019 I had completely forgotten  but which was reported at https://github.com/latex3/latex2e/issues/204 

If some Linux distro is using this specific LaTeX release, than its maintainers should be pinged to quickly move to some more recent update.

**Comment by fleimgruber** [comment]
@jfbu thanks for looking into this. The TL2019 was severely outdated and I could not find any mention of patch level in the `.log` file. I took this as a shove to upgrade to TL2020 which solved this for me.

Possibly relevant information for TL2019 from logs:

```
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019/W32TeX) (preloaded format=pdflatex 2019.10.7)  17 JAN 2021 12:05
entering extended mode
 restricted \write18 enabled.
 %&-line parsing enabled.
**manual.tex
(./manual.tex
LaTeX2e <2019-10-01>
(./sphinxhowto.cls
Document Class: sphinxhowto 2019/12/01 v2.3.0 Document class (Sphinx howto)
...
Package: graphicx 2017/06/01 v1.1a Enhanced LaTeX Graphics (DPC,SPQR)
...
```

please let me know if you need additional information for reproduction or confirmation. Otherwise this is fixed for me. @pique-b can you try with a more recent TL or give more information on your TL2019 installation with the infos from `.log` as mentioned in https://github.com/sphinx-doc/sphinx/issues/8135#issuecomment-761673721?

> If some Linux distro is using this specific LaTeX release, than its maintainers should be pinged to quickly move to some more recent update.

Good point but I guess such patch level fixes would be rolled out for the major distros as part of maintenance updates, right? More recent reports so far have been mostly (all?) on Windows.

**Comment by jfbu** [comment]
@fleimgruber Thanks for investigating. Yes the information you provide is all I need. I believe this issue is thus a duplicate of #6777, although it is complicated for me to reproduce:

- I do have a copy of TeXLive 2019 but it is the final one ; and I removed before archival all trace of the various updates during its life cycle,
- LaTeX has since a few years a rollback mechanism. I tried using it with manually adding ``\RequirePackage[2019-10-01]{latexrelease}`` but there seems to be no way to roll back to earlier than the final patch level. And the roll-back for `graphics` (called by `graphicx`) does not go back far enough. Although by digging into its sources I could try ``\RequirePackage{graphics}[=2017-06-25]``. This does break the build with some brace related error in `\sphinxincludegraphics`, but it is not quite the same error. Also, I don't know how to roll-back the `pdftex.def` driver code.
- I did have some left-over test of #6777, but only after LaTeX 2019-10-01 patch level 2 came out. Writing this I see perhaps I could fire up Time Machine and try to recover some log file with the failure as reported by @pique-b ... ah well it is so slow and anyway bad luck I can't find any archived log file from prior my final test with patch level 2 which solved it.

Anyway I am marking this as duplicate and will close after confirmation from @pique-b some more recent latex solves this.

As per



>     If some Linux distro is using this specific LaTeX release, than its maintainers should be pinged to quickly move to some more recent update.
> 
> Good point but I guess such patch level fixes would be rolled out for the major distros as part of maintenance updates, right? More recent reports so far have been mostly (all?) on Windows.

Let's hope so ! else there is my 

>  If syntax becomes usable again in a soon upcoming Patch Release, then this means production tools can consider not do anything and tell users in future "bad luck, update your LaTeX". 

from https://github.com/latex3/latex2e/issues/204 ;-)

**Comment by pique-b** [comment]
Thanks.
@jfbu: I am trying to dig out my old files (as since I moved to an unrelated task). Shall update with more info within a day.

@fleimgruber: As stated above in my original post, I am using Mac 10.12.6 and while trying to install the latest **Sphinx** package (so far as I recall), it failed to install as my **"os was of an older version and needed to upgrade it to Mac Sierra High"** (which my system hardware does not support). As stated, I am going to update (my post) with the steps taken by me to install Sphinx, LaTeX and other related packages within a day. In the mean time, could you please advise me the steps for upgrading from TL2019 to **TL2020** (and **the source**) so that I may try and run it once again.

**Comment by fleimgruber** [comment]
@pique-b oh, my bad. I am in a similar situation on my Mac using 10.12.6. MacTeX and TL2020 are only supported from 10.13 (High Sierra) onwards so we are out of luck I fear. You might find a TL mirror that still hosts the latest TL2019 and configure your system for that mirror (with the "TeX Live Utility"). See also http://tug.org/mactex/faq/3-4.html.

**Comment by pique-b** [comment]
@fleimgruber : From your 

> I took this as a shove to upgrade to TL2020 which solved this for me.

I deduced you might have cracked this (and hence my request for the **source**). Yeah, **_my bad, double over_**. In the meantime I am looking at the page you ref to.

**Comment by jfbu** [comment]
@pique-b @fleimgruber yes it is really a problem that the MacTeX (which is but a slim layer on TeXLive itself) installs only on the lest 3 years of Mac OS. But you can install TeXLive directly selecting the "darwin-legacy" platform. The instructions of direct TeXLive install are [there](https://tug.org/texlive/acquire-netinstall.html) (sorry that I could not find more user-friendly instruction page; basically you install on mac os exactly as on unixen, except that you select either the ``darwin`` or ``darwin-legacy`` platform depending how old your Mac OS is.).

In this eco-system there is no officially supported way from "upgrading" TeXLive from one yearly release to the next, one basically installs a full new 4 or 5 gigabytes thing (and decides whether to trash or keep the old install). There is a discouraged [way](https://tug.org/texlive/upgrade.html) to upgrade, only if you have been familiar with TeXLive for years and know how to navigate in your mac file system like on unix. I do the ``cp -a 2019 2020" step only because I usually want to keep a copy of previous year for testing and as remedy if some package gets broken, but it is possible to do that "in place". However, if you don't mind the multi-gigabytes download, and do not intend to keep the old install, it is easier to install the new yearly release, and to remove the old one. By the way, personnally I never install TeXLive as root user, because this has always caused me problems.

**Comment by jfbu** [comment]
@pique-b this being said, isn't it much easier if you simply fully upgrade your TeXLive 2019 (I suppose MacTeX) install ? Using TeXLiveUtility tool (I am not familiar since for many years I install TeXLive directly, not using MacTeX, and I use directly the command line tool ``tlmgr`` as on Linux distros to maintain my latex updated), there should be some interface to trigger full update of your installation. The upstream LaTeX problem was fixed during the year-life cycle of TeXLive 2019.

**Comment by jfbu** [comment]
@pique-b I could test with the latex release 2019-10-01 (prior to patch level 2) and reproduce at my locale the exact error you report



    LaTeX Warning: File `{img_file}.png' not found on input line 102.

    ! Package pdftex.def Error: File `"""{img_file}".png' not found: using draft se
    tting.

    See the pdftex.def package documentation for explanation.

(I got [help](https://github.com/latex3/latex2e/issues/204#issuecomment-761854532) at upstream latex on how to do that).

Somewhat optimistically I advised earlier to update your TeXLive 2019 to its final status. But indeed for this you need a server of "historic" versions of TeXLive as nowadays servers for updates will refuse to work with a TL2019 install. You can do that by telling TeXLive Utility (TLU) to use following mirror:

    http://ftp.math.utah.edu/pub/tex/historic/systems/texlive/2019/tlnet-final/

I tried this on my TeXLive 2019 (via command line interface tlmgr, as I don't use TLU) and it worked with this kind of output:

> $ tlmgr option repository "http://ftp.math.utah.edu/pub/tex/historic/systems/texlive/2019/tlnet-final"
> tlmgr: setting default package repository to http://ftp.math.utah.edu/pub/tex/historic/systems/texlive/2019/tlnet-final
> tlmgr: updating /usr/local/texlive/2019/tlpkg/texlive.tlpdb
> $ tlmgr update --list
> TeX Live 2019 is frozen forever and will no
> longer be updated.  This happens in preparation for a new release.
> 
> If you're interested in helping to pretest the new release (when
> pretests are available), please read https://tug.org/texlive/pretest.html.
> Otherwise, just wait, and the new release will be ready in due time.
> tlmgr: package repository http://ftp.math.utah.edu/pub/tex/historic/systems/texlive/2019/tlnet-final (not verified: valid signature with expired key)
> tlmgr: would save backups to /usr/local/texlive/2019/tlpkg/backups
> update:   tex4ht             [949k]: local:    53966, source:    54213
> 

This shows that my install was not completely final, package `tex4ht` had a last minute update before TL2019 was terminally frozen, that I did not have at my locale. Nothing was changed locally as I only ask for a list of things which would get updated.

Other historic servers are listed [here](https://tug.org/historic/).

I hope this will fix it for you. By the way, sorry for the very late answering to your bug report.

**Comment by fleimgruber** [comment]
@pique-b Just for reference, the steps I just went through to install a clean TL2019 and to upgrade to latest frozen using the GUI:

1. Download mactex-20190508.pkg from ftp://ftp.tug.org/historic/systems/mactex/2019
2. Install mactex-20190508.pkg
3. Open TeX Live Utility
4. Upgrade Open TeX Live Utility from 1.36 to 1.40
5. Restart TeX Live Utility
6. Point TeX Live Utility to http://ftp.math.utah.edu/pub/tex/historic/systems/texlive/2019/tlnet-final/ as mentioned in https://github.com/sphinx-doc/sphinx/issues/8135#issuecomment-761858861
6. Upgrade the texlive.infra.* (shown on Packages tab)
7. Restart TeX Live Utility
8. Update all packages

I had to do 7 and 8 twice, because of some hickups with ltxcmds.sty. The log should show

```
This is pdfTeX, Version 3.14159265-2.6-1.40.20 (TeX Live 2019) (preloaded format=pdflatex 2021.1.17)  17 JAN 2021 17:16
entering extended mode
 restricted \write18 enabled.
 %&-line parsing enabled.
**sphinx-issue-8135.tex
(./sphinx-issue-8135.tex
LaTeX2e <2020-02-02> patch level 5
```

and the issue should be fixed.

**Comment by pique-b** [comment]
@fleimgruber : I am a little confounded. My pdfTex version (**_as I have already narrated above_**) is as mentioned in your log i.e.

> pdfTeX 3.14159265-2.6-1.40.20 (TeX Live 2019)

Could you please let me know if there is a discrepancy? Or am I missing something?

_P.S._ Sorry for the late reply.

**Comment by jfbu** [comment]
@pique-b LaTeX uses an executable `pdftex` (or `xetex`, or  `luatex`, or `platex`) whose version varies rarely. But this executable will load whole bunch of pre-digested TeX macros, known as a `format`, e.g., `pdflatex` is (on unixen) a symlink to `pdftex` (more precisely `pdfetex`) and the latter knows (from the name under which it was called) that it has to use the ``pdflatex.fmt`` format file. This format file contains most of the ``latex2e`` source code. When a new LaTeX is released which happens twice a year, along with some extra patch releases, the format is "rebuilt" by the updating process; the ``pdftex`` executable itself never changes during the year-long life cycle of TeXLive/MacTeX. It changes only, if at all, from one year to the next. Also, the Sphinx created file will insert a whole lot of ``\usepackage{}`` things which load extra-code, not pre-digested into the format. So, in 99,999% of cases, bugs or problems have nothing to do with actual version number of the executable ``pdftex``. This is why the more important information is the LaTeX format version number, and in the version numbers of the various packages, as seen in the ``.log`` file of the latex build process.

It seems you have a TeXLive 2019 install, which however is stuck with a version of LaTeX whose `\includegraphics` does not accept braced basenames of graphics files. The fix is to update that. You do this via TeX Live Utility GUI program, but you have to use special server as in my previous comment and @fleimgruber's one. Possibly TexLive Utility will tell you it needs to update first its "infrastructure". At worst, you can start from a fresh install of MacTeX 2019 (as your mac os version is not supported by MacTeX 2020) but then you need to update it as @fleimgruber precisely explained. Hope it works for you!

**Comment by pique-b** [comment]
@jfbu / @fleimgruber :

Thanks for your responses. I am going to try **updating** first. And only if it does not help, try reinstalling MacTex (though I am afraid to, as in my first attempt I faced lot of  difficulties, and the whole process was, well, very discomforting what with many trial and errors and an unending set of support files having to install and "upgrade to").

Shall update on the outcome of my attempt/s by the weekend.

Thanks a lot.

**Comment by jfbu** [comment]
@pique-b good luck with this process of updating TL2019 which should go smoothly. TeXLive Utility should provide in its GUI interface an "update all packages" command. This will take some time though as hundreds of packages, besides the latex kernel files themselves, will get updated. When MacTeX still supported my mac os system, I recall I was using it that way but I have lost remembrance of how its menus looked like. Important thing is the server configuration as pointed out in previous comments. Else the process will stop immediately with a message that your install is not TL2020.

**Comment by jfbu** [comment]
I am closing this as being an upstream LaTeX problem which moreover was temporary during fall of 2019.
